src/mlpack/methods/decision_tree/gini_gain.hpp

```diff
@@ -33,26 +33,63 @@ class GiniGain
    *
    * @param labels Set of labels to evaluate Gini impurity on.
    * @param numClasses Number of classes in the dataset.
+   * @param weights Weight of labels.
    */
-  template<typename RowType>
+  template<bool UseWeights, typename RowType, typename WeightVecType>
   static double Evaluate(const RowType& labels,
-                         const size_t numClasses)
+                         const size_t numClasses,
+                         const WeightVecType& weights)
   {
     // Corner case: if there are no elements, the impurity is zero.
     if (labels.n_elem == 0)
       return 0.0;
 
-    arma::Col<size_t> counts(numClasses);
-    counts.zeros();
-    for (size_t i = 0; i < labels.n_elem; ++i)
-      counts[labels[i]]++;
+     // Count the number of elements in each class.
+     arma::Col<double> counts(numClasses);
+     counts.zeros();
 
     // Calculate the Gini impurity of the un-split node.
     double impurity = 0.0;
-    for (size_t i = 0; i < numClasses; ++i)
+
+    if (UseWeights)
+    {
+
+      // sum all the weights up
+      double accWeights = 0.0;
+
+      // the weight must have the same number of labels
+      if (weights.n_elem != labels.n_elem)
+      {
+        std::ostringstream oss;
+        oss << "the weights' length is' " << weights.n_elem << " but we have  " 
+            << labels.n_elem << " labels, but they must be equal";
+        throw std::invalid_argument(oss.str());
+      }
```

Comment:
Given that the `Evaluate()` function will be called so many times, I am fine with removing this check for speed (though it will be a minor boost if any) and simply writing in the documentation that the weight length must be the same as the label length (or the number of points).  The user is not likely to use this function directly anyway.  We can perform this check at a higher level in the code.